File: api/esi/esi.go
// Package esi is the Go implementation of a small subset of EVE's ESI API.
package esi

import (
	"errors"
	"net/http"

	resty "gopkg.in/resty.v1"
)

var endpoint = "https://esi.evetech.net/latest"

const (
	CategoryCharacter = "character"
)

// Entity ...
type Entity struct {
	ID       int    `json:"id"`
	Name     string `json:"name"`
	Category string `json:"category"`
}

// Entities ...
type Entities []*Entity

// FetchByID resolves a set of IDs to characters, corporations, allaiances, and more.
// https://esi.evetech.net/ui/#/Universe/post_universe_names
func (e *Entities) FetchByID(i []int) error {
	resp, err := resty.
		R().
		SetBody(i).
		SetResult(e).
		Post(endpoint + "/universe/names/")
	if err != nil {
		return err
	}
	if resp.StatusCode() != http.StatusOK {
		return errors.New("esi: " + resp.Status())
	}
	return nil
}

// FetchByName resolves a set of names to characters, corporations, allaiances, and more.
// https://esi.evetech.net/ui/#/Universe/post_universe_ids
func (e *Entities) FetchByName(n []string) error {
	data := map[string][]*Entity{}
	resp, err := resty.
		R().
		SetBody(n).
		SetResult(&data).
		Post(endpoint + "/universe/ids/")
	if err != nil {
		return err
	}
	if resp.StatusCode() != http.StatusOK {
		return errors.New("esi: " + resp.Status())
	}
	for category, entities := range data {
		for _, entity := range entities {
			switch category {
			case "characters":
				entity.Category = CategoryCharacter
			}
			*e = append(*e, entity)
		}
	}
	return nil
}

// Affiliation ...
type Affiliation struct {
	CharacterID   int `json:"character_id"`
	CorporationID int `json:"corporation_id"`
	AllianceID    int `json:"alliance_id"`
}

// Affiliations ...
type Affiliations []*Affiliation

// FetchByID resolves a set of character IDs to corporation, alliance and faction affiliations.
// https://esi.evetech.net/ui/#/Character/post_characters_affiliation
func (a *Affiliations) FetchByID(i []int) error {
	resp, err := resty.
		R().
		SetBody(i).
		SetResult(a).
		Post(endpoint + "/characters/affiliation/")
	if err != nil {
		return err
	}
	if resp.StatusCode() != http.StatusOK {
		return errors.New("esi: " + resp.Status())
	}
	return nil
}

File: client/src/api/esi.js
const baseUrl = "https://esi.evetech.net/latest";

async function request(path, options) {
  const response = await fetch(baseUrl + path, {
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json"
    },
    ...options
  });
  if (response.ok) {
    return response.json();
  }
  console.error(response);
  throw response;
}

export default {
  async fetchAffiliations(ids) {
    const json = await request(`/characters/affiliation/`, {
      method: "POST",
      body: JSON.stringify(ids)
    });
    return json.reduce((result, data) => {
      const affiliation = {
        corporationId: data.corporation_id
      };
      if (data.alliance_id) {
        affiliation.allianceId = data.alliance_id;
      }
      result[data.character_id] = affiliation;
      return result;
    }, {});
  },

  async fetchByIds(ids) {
    const json = await request("/universe/names/", {
      method: "POST",
      body: JSON.stringify(ids)
    });
    return json.reduce((result, data) => {
      if (result[data.id]) {
        console.log("ids are not unique between corporarion and alliance");
      }
      result[data.id] = data.name;
      return result;
    }, {});
  },

  async fetchByNames(names) {
    const json = await request("/universe/ids/", {
      method: "POST",
      body: JSON.stringify(names)
    });
    return json.characters.reduce((result, data) => {
      result[data.name] = data.id;
      return result;
    }, {});
  }
};
